pkg/skaffold/docker/authenticators.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+)
+
+var authenticators = Authenticators{}
+
+// Authenticators stores an authenticator per registry.
+type Authenticators struct {
+	byRegistry map[string]*lockedAuthenticator
+	lock       sync.Mutex
+}
+
+// For retrieves the authentiator for a given image reference.
+func (a *Authenticators) For(ref name.Reference) authn.Authenticator {
+	registry := ref.Context().Registry.Name()
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Get existing authenticator
+	if auth, present := a.byRegistry[registry]; present {
+		return auth
+	}
+
+	// Create a new authenticator
+	auth := &lockedAuthenticator{
+		delegate: a.newAuthenticator(ref),
+	}
+
+	if a.byRegistry == nil {
```

Comment:
if this is true, then could `a.lock` also be nil? (maybe this is the first time the object has been used since it was created)

Reply:
`sync.Mutex` would be nil indeed but that's ok. The nil value for `sync.Mutex` is valid and will self initialize in a safe manner. Magic!

Reply:
🌟